### PR TITLE
tool: constrain lsm visualizer b/w start and end version numbers

### PR DESCRIPTION
**Pull request by mufeez-amjad**
This change introduces 2 new CLI arguments for the lsm tool: `start-edit` and `end-edit`. 

These can be used to visualize only a contiguous range of manifest edits. I have also added an offset in the visualizer to make it explicit what the manifest edit range is.

image

Resolves https://github.com/cockroachdb/cockroach/issues/74537

Tested with [MANIFEST](https://github.com/cockroachdb/pebble/files/7857097/MANIFEST.sumeer-heuristics.zip)

### CODE REVIEW

**Review by bananabrick — COMMENTED** [review]
Nice work!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/pebble/1449#-:-MtEkzd1B08_-L4b4hrs:bdtx8r6)* status: 0 of 2 files reviewed, 4 unresolved discussions (waiting on @bananabrick and @mufeez-amjad)

___
*[-- commits, line 2 at r1](https://reviewable.io/reviews/cockroachdb/pebble/1449#-MtEoJBpD_Fvx-FCg6zM):*
We usually have commit messages have the same title and body as the pr. Something like `tool: constrain lsm visualizer b/w start and end version numbers` for the title, and the same body.
___
*[tool/lsm.go, line 55 at r1](https://reviewable.io/reviews/cockroachdb/pebble/1449#-MtElj0F2apx7JDyTiRs:-MtElj0F2apx7JDyTiRt:b-k1m8kr) ([raw file](https://github.com/cockroachdb/pebble/blob/0fdfe211f2ad2d4c4d3bec46c177c60468702fd2/tool/lsm.go#L55)):*
> ```Go
> 	Files     map[base.FileNum]lsmFileMetadata `json:",omitempty"`
> 	Keys      []lsmKey                         `json:",omitempty"`
> 	StartEdit int
> ```

I think this field is just being assigned, and never used.
___
*[tool/lsm.go, line 70 at r1](https://reviewable.io/reviews/cockroachdb/pebble/1449#-MtEl0yOEtHFMVTY2d5O:-MtEl0yOEtHFMVTY2d5P:b-txs1tc) ([raw file](https://github.com/cockroachdb/pebble/blob/0fdfe211f2ad2d4c4d3bec46c177c60468702fd2/tool/lsm.go#L70)):*
> ```Go
> 	startEdit int
> 	endEdit   int
> 	editCount int
> ```

Is there a problem with `startCount >= editCount`? Maybe return an error in that case.
___
*[tool/lsm.go, line 275 at r1](https://reviewable.io/reviews/cockroachdb/pebble/1449#-MtEmcirCBl4CnDHQDIq:-MtEmcirCBl4CnDHQDIr:b-8320kg) ([raw file](https://github.com/cockroachdb/pebble/blob/0fdfe211f2ad2d4c4d3bec46c177c60468702fd2/tool/lsm.go#L275)):*
> ```Go
> 			continue
> 		}
> 		v := manifest.NewVersion(l.cmp.Compare, l.fmtKey.fn, 0, currentFiles)
> ```

Maybe we can look at a cpu profile to make sure that the `NewVersion` function is indeed the most expensive part of this loop. That would assure us that the current method works. Would also get you used to interpreting cpu profiles, since we use those a lot.

**Review by bananabrick — COMMENTED** [review]
I gave this a whirl, and I couldn't reproduce the error which was occurring with the slider. 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/pebble/1449#-:-MtEwJP47yRoQbOikZZc:b3dzby6)* status: 0 of 2 files reviewed, 5 unresolved discussions (waiting on @bananabrick and @mufeez-amjad)

___
*[tool/lsm.go, line 69 at r1](https://reviewable.io/reviews/cockroachdb/pebble/1449#-MtEwZrsBpTaHcJFrMc5:-MtEwZrsBpTaHcJFrMc6:b-1p4ndq) ([raw file](https://github.com/cockroachdb/pebble/blob/0fdfe211f2ad2d4c4d3bec46c177c60468702fd2/tool/lsm.go#L69)):*
> ```Go
> 	pretty    bool
> 	startEdit int
> 	endEdit   int
> ```

Might want to return an error if `startEdit > endEdit`.

**Comment by mufeez-amjad** [comment]
### Benchmarks:

Ran both version of the tools using the above manifest with `--start-edit=1000 --end-edit=2000`

| Approach      | HTML Generation | DOMContentLoaded     |
| :---        |    :----:   |          ---: |
| [start,end] VE      | 3.15 user 0.24 system 0:02.57 real       | 10574.4 ms   |
|  [0, start) BVE + [start,end] VE | 2.16 user 0.15 system 0:01.84 real        | 720.3 ms      |

The performance of the BVE approach is better so I will continue with that approach.

File: tool/lsm.go
Comment (mufeez-amjad) [inline tool/lsm.go:246]: I'm not sure why the `ve.DeletedFiles` iteration was before the `ve.NewFiles` interation (or if it was necessary) but when changing the order the tool seems to function as normal. 

Changing the order helped avoid the logic of prepopulating `currentFiles` to include the BVE's new files (now they are added before `ve.DeletedFiles` are handled).

**Review by mufeez-amjad — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/pebble/1449)* status: 0 of 2 files reviewed, 5 unresolved discussions (waiting on @bananabrick)

___
*[-- commits, line 2 at r1](https://reviewable.io/reviews/cockroachdb/pebble/1449#-MtEoJBpD_Fvx-FCg6zM):*
<details><summary><i>Previously, bananabrick (Arjun Nair) wrote…</i></summary><blockquote>

We usually have commit messages have the same title and body as the pr. Something like `tool: constrain lsm visualizer b/w start and end version numbers` for the title, and the same body.
</blockquote></details>

Edited the commit message to match.
___
*[tool/lsm.go, line 55 at r1](https://reviewable.io/reviews/cockroachdb/pebble/1449#-MtElj0F2apx7JDyTiRs:-MtFdIEcDo-m7VlLEoHJ:b-mlp7zq) ([raw file](https://github.com/cockroachdb/pebble/blob/0fdfe211f2ad2d4c4d3bec46c177c60468702fd2/tool/lsm.go#L55)):*
<details><summary><i>Previously, bananabrick (Arjun Nair) wrote…</i></summary><blockquote>

I think this field is just being assigned, and never used.
</blockquote></details>

It's not used on the go side, but it's used to offset the slider scale in the webpage's js.
 
___
*[tool/lsm.go, line 69 at r1](https://reviewable.io/reviews/cockroachdb/pebble/1449#-MtEwZrsBpTaHcJFrMc5:-MtKOxjhCpofvedUVZBY:b-896fix) ([raw file](https://github.com/cockroachdb/pebble/blob/0fdfe211f2ad2d4c4d3bec46c177c60468702fd2/tool/lsm.go#L69)):*
<details><summary><i>Previously, bananabrick (Arjun Nair) wrote…</i></summary><blockquote>

Might want to return an error if `startEdit > endEdit`.
</blockquote></details>

Done.
___
*[tool/lsm.go, line 70 at r1](https://reviewable.io/reviews/cockroachdb/pebble/1449#-MtEl0yOEtHFMVTY2d5O:-MtKOyv6-tf8X-XuyEAJ:b-896fix) ([raw file](https://github.com/cockroachdb/pebble/blob/0fdfe211f2ad2d4c4d3bec46c177c60468702fd2/tool/lsm.go#L70)):*
<details><summary><i>Previously, bananabrick (Arjun Nair) wrote…</i></summary><blockquote>

Is there a problem with `startCount >= editCount`? Maybe return an error in that case.
</blockquote></details>

Done.
___
*[tool/lsm.go, line 275 at r1](https://reviewable.io/reviews/cockroachdb/pebble/1449#-MtEmcirCBl4CnDHQDIq:-MtK_cWKE6GhGHyCJtrC:bdgu47q) ([raw file](https://github.com/cockroachdb/pebble/blob/0fdfe211f2ad2d4c4d3bec46c177c60468702fd2/tool/lsm.go#L275)):*
<details><summary><i>Previously, bananabrick (Arjun Nair) wrote…</i></summary><blockquote>

Maybe we can look at a cpu profile to make sure that the `NewVersion` function is indeed the most expensive part of this loop. That would assure us that the current method works. Would also get you used to interpreting cpu profiles, since we use those a lot.
</blockquote></details>

Skipping for this issue.

**Review by bananabrick — COMMENTED** [review]
Nice work figuring out a faster method!

Do you have any thoughts on why this method is faster? I don't see too much of a difference. In both cases, we iterate over the files added/deleted, over all version edits from [0, start). It almost seems like more work, because we iterate over the added/deleted files twice, once in `bve.Accumulate`, and once after that to add the files to the `startingEdit`.

Did you also use the version of the tool on master, your previous version, this version, to verify that the result produced in the graph for a given range of version edits is identical?

I think there's something off here, although there's a chance I ran this incorrectly. For example, try running:
`./pebble lsm --start-edit=1 --end-edit=2 MANI > lsm.html` where `MANI` is just the manifest file you were using for testing.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/pebble/1449#-:-MtOPaUi3uJby1n-Ldvz:b-xcqkyp)* status: 0 of 2 files reviewed, 7 unresolved discussions (waiting on @bananabrick and @mufeez-amjad)

___
*[tool/lsm.go, line 126 at r7](https://reviewable.io/reviews/cockroachdb/pebble/1449#-MtOWZ4Y7l5Sv_NuyPd-:-MtOWZ4Y7l5Sv_NuyPd0:b-9h8ws0) ([raw file](https://github.com/cockroachdb/pebble/blob/c2fbfe2e8b40cf8a335c75ac959fc04543708cc4/tool/lsm.go#L126)):*
> ```Go
> 		}
> 	}
> ```

Might be worthwhile to add another check for `startEdit >= len(edits)`.
___
*[tool/lsm.go, line 127 at r7](https://reviewable.io/reviews/cockroachdb/pebble/1449#-MtOR4K-2ZtosgbHddTz:-MtOR4K-2ZtosgbHddU-:b-92otu) ([raw file](https://github.com/cockroachdb/pebble/blob/c2fbfe2e8b40cf8a335c75ac959fc04543708cc4/tool/lsm.go#L127)):*
> ```Go
> 	}
> 
> 	if l.startEdit >= l.endEdit {
> ```

nit: Should `startEdit == endEdit` be allowed?
___
*[tool/lsm.go, line 152 at r7](https://reviewable.io/reviews/cockroachdb/pebble/1449#-MtOTGRp6pNuK0sNXZt0:-MtOTGRp6pNuK0sNXZt1:bwv53b4) ([raw file](https://github.com/cockroachdb/pebble/blob/c2fbfe2e8b40cf8a335c75ac959fc04543708cc4/tool/lsm.go#L152)):*
> ```Go
> 	}
> 	if l.endEdit < len(edits) {
> 		edits = edits[:l.endEdit-l.startEdit]
> ```

nit: Is `endEdit` not inclusive? If `startEdit` is `0`, and `endEdit` is `1`,  this would make edits have `len` of  `1`.
___
*[tool/lsm.go, line 339 at r7](https://reviewable.io/reviews/cockroachdb/pebble/1449#-MtOWAcH2DVuZzGdVTmE:-MtOWAcH2DVuZzGdVTmF:b3eqwxf) ([raw file](https://github.com/cockroachdb/pebble/blob/c2fbfe2e8b40cf8a335c75ac959fc04543708cc4/tool/lsm.go#L339)):*
> ```Go
> 	}
> 
> 	startingEdit := &edits[l.startEdit]
> ```

Why is `startingEdit` a pointer to a pointer?
___
*[tool/lsm.go, line 342 at r7](https://reviewable.io/reviews/cockroachdb/pebble/1449#-MtOdEvUDVf10qyf7Ttw:-MtOdEvUDVf10qyf7Ttx:b4fi33k) ([raw file](https://github.com/cockroachdb/pebble/blob/c2fbfe2e8b40cf8a335c75ac959fc04543708cc4/tool/lsm.go#L342)):*
> ```Go
> 	var beNewFiles []manifest.NewFileEntry
> 
> 	for level, deletedFiles := range be.Deleted {
> ```

For me, mutating the `versionEdit` at `startEdit`, to include all the file information from `[0, startEdit)` is confusing. Instead, I would pre-populate whatever data you need and pass that into `buildKeys`, `buildEdits`, and keep `startEdit` the way it is. Although I'm happy to be convinced that the current method is better/required.
___
*[tool/lsm.go, line 368 at r7](https://reviewable.io/reviews/cockroachdb/pebble/1449#-MtOexqq-X0U_nuF1Zjv:-MtOexqq-X0U_nuF1Zjw:b2nu2o6) ([raw file](https://github.com/cockroachdb/pebble/blob/c2fbfe2e8b40cf8a335c75ac959fc04543708cc4/tool/lsm.go#L368)):*
> ```Go
> 	}
> 	(*startingEdit).NewFiles = append(beNewFiles, (*startingEdit).NewFiles...)
> 	(*startingEdit).DeletedFiles = nil
> ```

Is setting `startingEdit.DeletedFiles = nil` correct?
___
*[tool/lsm_data.go, line 316 at r7](https://reviewable.io/reviews/cockroachdb/pebble/1449#-MtOhtqC-Zz9J6i610HM:-MtOhtqC-Zz9J6i610HN:bkgceb5) ([raw file](https://github.com/cockroachdb/pebble/blob/c2fbfe2e8b40cf8a335c75ac959fc04543708cc4/tool/lsm_data.go#L316)):*
> ```Go
>             let sublevel = null;
>             for (let i = index; i >= 0 && (sublevel === null || sublevel === undefined); i--) {
>                 if (!data.Edits[i].Sublevels) {
> ```

Why was this not required before, and why is it required now?

**Review by bananabrick — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/pebble/1449)* status: 0 of 2 files reviewed, 7 unresolved discussions (waiting on @bananabrick and @mufeez-amjad)

___
*[tool/lsm_data.go, line 750 at r2](https://reviewable.io/reviews/cockroachdb/pebble/1449#-MtPPKkN228nz73VUJlO:-MtPPKkOCH4Vhsk0F6I4:b-n3f008) ([raw file](https://github.com/cockroachdb/pebble/blob/e6d9680ed428c737d615065d3e64fb3f664e5013/tool/lsm_data.go#L750)):*
> ```Go
>             .domain([0, data.Edits.length - 1])
>             .range([0, width])
>             .clamp(true);
> ```

Btw, I just realized that this file is generated and shouldn't be edited. See, `make_lsm_data.sh`.

**Review by mufeez-amjad — COMMENTED** [review]
> Do you have any thoughts on why this method is faster? 

I think most of the improvement comes from the significant decrease in amount of data that's needed to be written to the HTML file (and thus decrease the time it takes for page load). Because we coalesce `[0, startEdit)`, there's no need to keep data on files and keys outside of `[startEdit, endEdit]`.

> Did you also use the version of the tool on master, your previous version, this version, to verify that the result produced in the graph for a given range of version edits is identical?

I ran the manifest using `--edit-count=2000` on both master and my latest commit and the result is identical.


I also ran `start-edit=1000 end-edit=2000` and the result is almost identical (my changes have an empty L0.13 sublevel missing which is present only in the original because that sublevel is only occupied in `[0, startEdit)` and not in `[startEdit, endEdit]`).

.png)

Note that the difference in index (1000 vs 1002 is because the first 2 edits of the manifest are not included because New/Deleted files are empty, but with my changes it is 1002 because that is startEdit + 2 (empty edits).

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/pebble/1449#-:-MtPCVomBfgwrU2uc20b:b-2hrw12)* status: 0 of 2 files reviewed, 7 unresolved discussions (waiting on @bananabrick and @mufeez-amjad)

___
*[tool/lsm.go, line 126 at r7](https://reviewable.io/reviews/cockroachdb/pebble/1449#-MtOWZ4Y7l5Sv_NuyPd-:-MtPEntW7iyPegNoIHfz:b-896fix) ([raw file](https://github.com/cockroachdb/pebble/blob/c2fbfe2e8b40cf8a335c75ac959fc04543708cc4/tool/lsm.go#L126)):*
<details><summary><i>Previously, bananabrick (Arjun Nair) wrote…</i></summary><blockquote>

Might be worthwhile to add another check for `startEdit >= len(edits)`.
</blockquote></details>

Done.
___
*[tool/lsm.go, line 127 at r7](https://reviewable.io/reviews/cockroachdb/pebble/1449#-MtOR4K-2ZtosgbHddTz:-MtPU2P6A15b1Vo400Hy:bdpwoh0) ([raw file](https://github.com/cockroachdb/pebble/blob/c2fbfe2e8b40cf8a335c75ac959fc04543708cc4/tool/lsm.go#L127)):*
<details><summary><i>Previously, bananabrick (Arjun Nair) wrote…</i></summary><blockquote>

nit: Should `startEdit == endEdit` be allowed?
</blockquote></details>

With the inclusive `endEdit` change, I think it should. However, one issue is that for something like `start-edit=1 end-edit=1` there are no added/deleted files so data.Edits is undefined in the js - I added a check and return an error if this is the case.
___
*[tool/lsm.go, line 152 at r7](https://reviewable.io/reviews/cockroachdb/pebble/1449#-MtOTGRp6pNuK0sNXZt0:-MtPTA0Y56q21ORYgA-9:b-rif639) ([raw file](https://github.com/cockroachdb/pebble/blob/c2fbfe2e8b40cf8a335c75ac959fc04543708cc4/tool/lsm.go#L152)):*
<details><summary><i>Previously, bananabrick (Arjun Nair) wrote…</i></summary><blockquote>

nit: Is `endEdit` not inclusive? If `startEdit` is `0`, and `endEdit` is `1`,  this would make edits have `len` of  `1`.
</blockquote></details>

Fixed to be inclusive.
___
*[tool/lsm.go, line 339 at r7](https://reviewable.io/reviews/cockroachdb/pebble/1449#-MtOWAcH2DVuZzGdVTmE:-MtPT-rB7qeGwdJG4uHh:b-l97i80) ([raw file](https://github.com/cockroachdb/pebble/blob/c2fbfe2e8b40cf8a335c75ac959fc04543708cc4/tool/lsm.go#L339)):*
<details><summary><i>Previously, bananabrick (Arjun Nair) wrote…</i></summary><blockquote>

Why is `startingEdit` a pointer to a pointer?
</blockquote></details>

Good point, fixed.
___
*[tool/lsm.go, line 342 at r7](https://reviewable.io/reviews/cockroachdb/pebble/1449#-MtOdEvUDVf10qyf7Ttw:-MtPUCCLBfG5jdtvvjpA:bwbqm6f) ([raw file](https://github.com/cockroachdb/pebble/blob/c2fbfe2e8b40cf8a335c75ac959fc04543708cc4/tool/lsm.go#L342)):*
<details><summary><i>Previously, bananabrick (Arjun Nair) wrote…</i></summary><blockquote>

For me, mutating the `versionEdit` at `startEdit`, to include all the file information from `[0, startEdit)` is confusing. Instead, I would pre-populate whatever data you need and pass that into `buildKeys`, `buildEdits`, and keep `startEdit` the way it is. Although I'm happy to be convinced that the current method is better/required.
</blockquote></details>

The motivation for the approach I took was that the startEdit is the starting point of the visualization, and so it should account for all previous additions/deletions to the LSM (the previous approach did not account for them). It seems like what you're suggesting with pre-populated data would require special handling of `edits[0]` in `buildEdits` and would bloat the for loop.
___
*[tool/lsm.go, line 368 at r7](https://reviewable.io/reviews/cockroachdb/pebble/1449#-MtOexqq-X0U_nuF1Zjv:-MtPF_8x4n3q_fWshrw1:b-5gho2m) ([raw file](https://github.com/cockroachdb/pebble/blob/c2fbfe2e8b40cf8a335c75ac959fc04543708cc4/tool/lsm.go#L368)):*
<details><summary><i>Previously, bananabrick (Arjun Nair) wrote…</i></summary><blockquote>

Is setting `startingEdit.DeletedFiles = nil` correct?
</blockquote></details>

This is part of the optimization to not include unnecessary data before `startingEdit`. `startingEdit` can delete files from `[0, startEdit)` but in the js tool, that would require having a record of the deleted files in `l.state.Files` (`data.Files`). Instead, I remove the deleted files from `startingEdit.NewFiles` and mark `startingEdit.DeletedFiles` as null to remove dangling references.

**Review by bananabrick — COMMENTED** [review]
Thanks for the explanations. Reminder to squash commits into one before merging.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/pebble/1449#-:-MtcuMpuC81lY6g20mmT:bqx4aga)* status: 0 of 3 files reviewed, 4 unresolved discussions (waiting on @bananabrick and @mufeez-amjad)

___
*[tool/lsm.go, line 368 at r7](https://reviewable.io/reviews/cockroachdb/pebble/1449#-MtOexqq-X0U_nuF1Zjv:-Mte4mAR5TI0Q3IG-1LQ:bh91uzv) ([raw file](https://github.com/cockroachdb/pebble/blob/c2fbfe2e8b40cf8a335c75ac959fc04543708cc4/tool/lsm.go#L368)):*
<details><summary><i>Previously, mufeez-amjad (Mufeez Amjad) wrote…</i></summary><blockquote>

This is part of the optimization to not include unnecessary data before `startingEdit`. `startingEdit` can delete files from `[0, startEdit)` but in the js tool, that would require having a record of the deleted files in `l.state.Files` (`data.Files`). Instead, I remove the deleted files from `startingEdit.NewFiles` and mark `startingEdit.DeletedFiles` as null to remove dangling references. 
</blockquote></details>

`startingEdit` can also delete a file which was added at `startingEdit`. The current change doesn't modify `startingEdit.NewFiles`, and adds it to the `beNewFiles`, and turns that into the new `startingEdit.NewFiles`. Which means we could  miss some deleted files at `startingEdit.DeletedFiles`. Is that correct?
___
*[tool/lsm.go, line 330 at r9](https://reviewable.io/reviews/cockroachdb/pebble/1449#-Mte5nAp0KZHethYeU2r:-Mte5nAp0KZHethYeU2s:bs5yl0v) ([raw file](https://github.com/cockroachdb/pebble/blob/06b137b471dc840373f0f9b6e6fd0bc72f6dccb7/tool/lsm.go#L330)):*
> ```Go
> 	}
> 
> 	if l.state.Edits == nil {
> ```

If `len(edits) > 0`, will this condition ever be true? In the loop, we only `continue`, if `len(ve.DeletedFiles) == 0 && len(ve.NewFiles) == 0`, but I don't expect this condition to be true for a versionEdit. Did the manifest you were testing with have a `versionEdit` like this?
___
*[tool/data/lsm.js, line 207 at r9](https://reviewable.io/reviews/cockroachdb/pebble/1449#-Mte6WLQ2pMnHnB1Ybzh:-Mte6WLQ2pMnHnB1Ybzi:b-3vdfia) ([raw file](https://github.com/cockroachdb/pebble/blob/06b137b471dc840373f0f9b6e6fd0bc72f6dccb7/tool/data/lsm.js#L207)):*
> ```JavaScript
>         for (let file of this.levels[0]) {
>             let sublevel = null;
>             for (let i = index; i >= 0 && (sublevel === null || sublevel === undefined) && data.Edits[i].Sublevels; i--) {
> ```

Looks like we set `Edits[i].Sublevels` for every single edit, so why is this condition required?

**Review by bananabrick — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/pebble/1449)* status: 0 of 3 files reviewed, 4 unresolved discussions (waiting on @bananabrick and @mufeez-amjad)

___
*[tool/lsm.go, line 330 at r9](https://reviewable.io/reviews/cockroachdb/pebble/1449#-Mte5nAp0KZHethYeU2r:-MteAp222dGzJj0Gwxzp:bu3zxbp) ([raw file](https://github.com/cockroachdb/pebble/blob/06b137b471dc840373f0f9b6e6fd0bc72f6dccb7/tool/lsm.go#L330)):*
<details><summary><i>Previously, bananabrick (Arjun Nair) wrote…</i></summary><blockquote>

If `len(edits) > 0`, will this condition ever be true? In the loop, we only `continue`, if `len(ve.DeletedFiles) == 0 && len(ve.NewFiles) == 0`, but I don't expect this condition to be true for a versionEdit. Did the manifest you were testing with have a `versionEdit` like this?
</blockquote></details>

Discussed this offline. I would change this to, "there are no edits in [start-edit, end-edit], which add or delete files.

**Review by mufeez-amjad — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/pebble/1449)* status: 0 of 3 files reviewed, 4 unresolved discussions (waiting on @bananabrick)

___
*[tool/lsm.go, line 368 at r7](https://reviewable.io/reviews/cockroachdb/pebble/1449#-MtOexqq-X0U_nuF1Zjv:-MteBmCW86885obp_b23:b82ogag) ([raw file](https://github.com/cockroachdb/pebble/blob/c2fbfe2e8b40cf8a335c75ac959fc04543708cc4/tool/lsm.go#L368)):*
<details><summary><i>Previously, bananabrick (Arjun Nair) wrote…</i></summary><blockquote>

`startingEdit` can also delete a file which was added at `startingEdit`. The current change doesn't modify `startingEdit.NewFiles`, and adds it to the `beNewFiles`, and turns that into the new `startingEdit.NewFiles`. Which means we could  miss some deleted files at `startingEdit.DeletedFiles`. Is that correct?
</blockquote></details>

Good catch, I didn't know this was an edge case to consider. Fixed to leave `startingEdit.DeletedFiles` unmodified.
___
*[tool/lsm.go, line 330 at r9](https://reviewable.io/reviews/cockroachdb/pebble/1449#-Mte5nAp0KZHethYeU2r:-MteBXiL34AMqOxUrfXM:b-896fix) ([raw file](https://github.com/cockroachdb/pebble/blob/06b137b471dc840373f0f9b6e6fd0bc72f6dccb7/tool/lsm.go#L330)):*
<details><summary><i>Previously, bananabrick (Arjun Nair) wrote…</i></summary><blockquote>

Discussed this offline. I would change this to, "there are no edits in [start-edit, end-edit], which add or delete files.
</blockquote></details>

Done.
___
*[tool/lsm_data.go, line 316 at r7](https://reviewable.io/reviews/cockroachdb/pebble/1449#-MtOhtqC-Zz9J6i610HM:-MteSoZgDJIb-qY4fU4-:b90g3qm) ([raw file](https://github.com/cockroachdb/pebble/blob/c2fbfe2e8b40cf8a335c75ac959fc04543708cc4/tool/lsm_data.go#L316)):*
<details><summary><i>Previously, bananabrick (Arjun Nair) wrote…</i></summary><blockquote>

Why was this not required before, and why is it required now?
</blockquote></details>

Responded below.
___
*[tool/data/lsm.js, line 207 at r9](https://reviewable.io/reviews/cockroachdb/pebble/1449#-Mte6WLQ2pMnHnB1Ybzh:-MteSptM567-3rnJTELw:b-vk1d3f) ([raw file](https://github.com/cockroachdb/pebble/blob/06b137b471dc840373f0f9b6e6fd0bc72f6dccb7/tool/data/lsm.js#L207)):*
<details><summary><i>Previously, bananabrick (Arjun Nair) wrote…</i></summary><blockquote>

Looks like we set `Edits[i].Sublevels` for every single edit, so why is this condition required?
</blockquote></details>

If there are no L0 sublevel changes between edits, `edit.Sublevels` is empty and because of `omitempty` in the struct declaration, the json encoded format doesn't include `Sublevels` for that edit. I spoke to Bilal and skipping any undefined map is acceptable.

**Review by bananabrick — APPROVED** [review]
> Reviewed 3 of 3 files at r10, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/pebble/1449)* status: all files reviewed, 1 unresolved discussion (waiting on @mufeez-amjad)

___
*[tool/data/lsm.js, line 207 at r9](https://reviewable.io/reviews/cockroachdb/pebble/1449#-Mte6WLQ2pMnHnB1Ybzh:-Mtetd9JAGMXe6bVJ0Cw:b-mg93x6) ([raw file](https://github.com/cockroachdb/pebble/blob/06b137b471dc840373f0f9b6e6fd0bc72f6dccb7/tool/data/lsm.js#L207)):*
<details><summary><i>Previously, mufeez-amjad (Mufeez Amjad) wrote…</i></summary><blockquote>

If there are no L0 sublevel changes between edits, `edit.Sublevels` is empty and because of `omitempty` in the struct declaration, the json encoded format doesn't include `Sublevels` for that edit. I spoke to Bilal and skipping any undefined map is acceptable.
</blockquote></details>

Nice. I think the `continue` looks good.

**Comment by bananabrick** [comment]
For Pebble, you can just hit the rebase and merge button.

**Pull request merged**
